action/action.go

```diff
@@ -29,6 +31,122 @@ type Action interface {
 	Run(*claim.Claim, credentials.Set, io.Writer) error
 }
 
+func golangTypeToJSONType(value interface{}) (string, error) {
+	switch v := value.(type) {
+	case nil:
+		return "null", nil
+	case bool:
+		return "boolean", nil
+	case float64:
+		// All numeric values are parsed by JSON into float64s. When a value could be an integer, it could also be a number, so give the more specific answer.
+		if math.Trunc(v) == v {
+			return "integer", nil
+		}
+		return "number", nil
+	case string:
+		return "string", nil
+	case map[string]interface{}:
+		return "object", nil
+	case []interface{}:
+		return "array", nil
+	default:
+		return fmt.Sprintf("%T", value), fmt.Errorf("unsupported type: %T", value)
+	}
+}
+
+// allowedTypes takes an output Schema and returns a map of the allowed types (to true)
+// and a string that is a list of all the allowed types (for error messagse)
+// or an error (if reading the allowed types from the schema failed).
+func allowedTypes(outputSchema definition.Schema) (map[string]bool, string, error) {
```

Comment:
What makes the `map[string]struct{}` representation idiomatic?

It looks like [_Effective Go_ approves of `map[string]bool`](https://golang.org/doc/effective_go.html#maps).

Reply:
It's just common practice because `struct{}` occupies 0 bytes unlike `bool`. Also, the `map[string]bool` representation isn't ideal because there are three possibilities for any given string (absent, present and mapping to `true`, present and mapping to `false`) which isn't what you really want for a set.

I'm disappointed that Effective Go hasn't picked up on this.

Reply:
I'm not worried about the memory allocation here. This is a set of allowed types names; there's a small a constant number of JSON types, so in practice this will be tiny.

Reply:
I agree. I think the representation issue is more important, but I don't want to argue and it will be easy to switch the types later.

Reply:
For the record, Dave Cheney also [prefers](https://dave.cheney.net/2014/03/25/the-empty-struct) `map[T]bool` as way of representing a set of `T`. I still prefer `map[T]struct{}`, but it's only a preference.